### PR TITLE
Relax isValidBucketName to allow reading existing buckets.

### PR DESCRIPTION
**Pull request by harshavardhana**
Add a stricter check for MakeBucket alone to avoid
creating newer buckets with capitalized letters.

Fixes https://github.com/minio/mc/issues/2157

### CODE REVIEW

**Comment by harshavardhana** [comment]
> isValidObjectName()
> isValidObjectNameStrict()
> 
> isValidBucketName()
> isValidBucketNameStrict()
> isValidBucketNameCommon()

Yes this is how it was in one of the versions, added an interface to see if its easier to bring a generic validator which can be initialized and used.

Why do you think interface is a problem? @krishnasrinivas

**Comment by krishnasrinivas** [comment]
> Why do you think interface is a problem? @krishnasrinivas

It is compromising code simplicity here.

Notice that this commit has introduced:
3 new types: Validator, BucketNameValidator, ObjectNameValidator
2 global vars: bucketNameValidator, objectNameValidator
5 methods: checkCommon, 2 x Check(), 2 x CheckStrict()
4 functions: isValidBucketName, isValidBucketNameStrict, isValidObjectPrefix, isValidObjectName

Instead of this we could just have:
5 functions: isValidBucketName, isValidBucketNameStrict, isValidBucketNameCommon, isValidObjectPrefix, isValidObjectName

File: pkg/s3utils/utils.go [hunk 1]
```diff
@@ -181,3 +182,126 @@ func EncodePath(pathName string) string {
...
+			default:
+				return errors.New("Bucket name contains invalid characters")
+			}
+		}
+		allNumbers = allNumbers && !isNotNumber
+	}
+	// Does the bucket name look like an IP address?
+	if len(pieces) == 4 && allNumbers {
+		return errors.New("Bucket name cannot be an ip address")
+	}
```
Comment (krishnasrinivas) [inline pkg/s3utils/utils.go]: Why not use regex and let it do all the heavy lifting?

Comment (harshavardhana) [inline pkg/s3utils/utils.go]: Well this is copied from server code @krishnasrinivas . We can keep the regex no big deal i think.

**Comment by harshavardhana** [comment]
> Notice that this commit has introduced:
> 3 new types: Validator, BucketNameValidator, ObjectNameValidator
> 2 global vars: bucketNameValidator, objectNameValidator
> 5 methods: checkCommon, 2 x Check(), 2 x CheckStrict()
> 4 functions: isValidBucketName, isValidBucketNameStrict, isValidObjectPrefix, isValidObjectName
> 
> Instead of this we could just have:
> 5 functions: isValidBucketName, isValidBucketNameStrict, isValidBucketNameCommon, isValidObjectPrefix, isValidObjectName

Just because we added more functions doesn't mean that code is not simple.. but i will get back the other changes i had stashed before interface. :-)

**Comment by krishnasrinivas** [comment]
>> It is compromising code simplicity here.

> Just because we added more functions doesn't mean that code is not simple

I meant the code in this commit is relatively more complex than if you were to just use functions. i.e if you have the two implementations side by side, which one is simpler.

File: utils.go [hunk 2]
```diff
@@ -115,6 +113,22 @@ var emptySHA256 = sum256(nil)
...
+func isValidBucketNameStrict(bucketName string) error {
+	return s3utils.CheckValidBucketNameStrict(bucketName)
+}
+
+func isValidObjectPrefix(objectPrefix string) error {
+	return s3utils.CheckValidObjectName(objectPrefix)
+}
+
+func isValidObjectName(objectName string) error {
+	return s3utils.CheckValidBucketNameStrict(objectName)
```
Comment (vadmeste) [inline utils.go]: This supposed to be s3utils.CheckValidObjectNameStrict(objectName)

Comment (harshavardhana) [inline utils.go]: Oh yes..oops.

**Comment by krishnasrinivas** [comment]
@harshavardhana what is the convention used on when CheckValidObjectNameStrict and CheckValidObjectName is used? Because GetObject uses CheckValidObjectNameStrict. But ListObjects uses CheckValidObjectName for prefix validation. Also BucketPolicy calls use CheckValidObjectName

**Comment by harshavardhana** [comment]
> @harshavardhana what is the convention used on when CheckValidObjectNameStrict and CheckValidObjectName is used? Because GetObject uses CheckValidObjectNameStrict. But ListObjects uses CheckValidObjectName for prefix validation. Also BucketPolicy calls use CheckValidObjectName

The reason is one is a prefix which can be empty as well and another is a whole object name. But it is just not clear.. I think let me make it clear rather than calling it strict.

File: api-put-bucket.go [hunk 3]
```diff
@@ -118,10 +108,10 @@ func (c Client) MakeBucket(bucketName string, location string) (err error) {
 //  writeonly - anonymous put/delete access to a given object prefix.
 func (c Client) SetBucketPolicy(bucketName string, objectPrefix string, bucketPolicy policy.BucketPolicy) error {
 	// Input validation.
-	if err := isValidBucketName(bucketName); err != nil {
+	if err := s3utils.CheckValidBucketName(bucketName); err != nil {
 		return err
 	}
-	if err := isValidObjectPrefix(objectPrefix); err != nil {
+	if err := s3utils.CheckValidObjectName(objectPrefix); err != nil {
```
Comment (vadmeste) [inline api-put-bucket.go]: This should be `s3utils.CheckValidObjectNamePrefix`

**Comment by harshavardhana** [comment]
@vadmeste @krishnasrinivas the PR is updated and rebased.. Can you guys take a look?

**Pull request merged**